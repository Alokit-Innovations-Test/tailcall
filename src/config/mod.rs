mod config;
mod from_document;
pub mod group_by;
mod into_document;
mod key_values;
mod n_plus_one;
mod server;
mod source;
mod js_plugin;

pub use config::*;
pub use key_values::*;
pub use server::*;
pub use source::*;
pub use js_plugin::*;

fn is_default<T: Default + Eq>(val: &T) -> bool {
  *val == T::default()
}
