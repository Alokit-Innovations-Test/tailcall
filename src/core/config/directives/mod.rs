use lazy_static::lazy_static;

mod add_field;
mod alias;
mod cache;
mod call;
mod expr;
mod federation;
mod graphql;
mod grpc;
mod http;
mod js;
mod link;
mod modify;
mod omit;
mod protected;
mod server;
mod telemetry;
mod upstream;

pub use add_field::*;
pub use alias::*;
pub use cache::*;
pub use call::*;
pub use expr::*;
pub use federation::*;
pub use graphql::*;
pub use grpc::*;
pub use http::*;
pub use js::*;
pub use link::*;
pub use modify::*;
pub use omit::*;
pub use protected::*;
pub use server::*;
pub use telemetry::*;
pub use upstream::*;

use crate::core::directive::DirectiveCodec;

lazy_static! {
    pub static ref KNOWN_DIRECTIVES: Vec<String> = vec![
        add_field::AddField::directive_name(),
        alias::Alias::directive_name(),
        cache::Cache::directive_name(),
        call::Call::directive_name(),
        expr::Expr::directive_name(),
        graphql::GraphQL::directive_name(),
        grpc::Grpc::directive_name(),
        http::Http::directive_name(),
        js::JS::directive_name(),
        link::Link::directive_name(),
        modify::Modify::directive_name(),
        omit::Omit::directive_name(),
        protected::Protected::directive_name(),
        server::Server::directive_name(),
        telemetry::Telemetry::directive_name(),
        upstream::Upstream::directive_name(),
    ];
}
