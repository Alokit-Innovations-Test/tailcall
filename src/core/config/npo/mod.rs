mod identifier;
mod queries;

use std::{collections::HashSet, rc::Rc};

pub use identifier::*;
pub use queries::*;

use super::Config;

struct NPOIdentifier<'a> {
    config: &'a Config,
}

impl<'a> NPOIdentifier<'a> {
    pub fn new(config: &'a Config) -> NPOIdentifier {
        NPOIdentifier { config }
    }

    fn iter(
        &self,
        path: Chunk<FieldName<'a>>,
        type_name: TypeName<'a>,
        is_list: bool,
        visited: &mut HashSet<TypeName<'a>>,
    ) -> Chunk<Chunk<FieldName<'a>>> {
        if visited.contains(&type_name) {
            return Chunk::new();
        } else {
            visited.insert(type_name);
        }
        let mut chunks = Chunk::new();
        if let Some(type_of) = self.config.find_type(&type_name.as_str()) {
            for (name, field) in type_of.fields.iter() {
                let path = path.clone().append(FieldName::new(name));
                let is_batch = field.has_batched_resolver();

                if field.has_resolver() {
                    if !is_batch && is_list {
                        chunks = chunks.append(path.clone());
                    }
                }

                let is_list = is_list | field.list;

                chunks = chunks.concat(self.iter(
                    path,
                    TypeName::new(field.type_of.as_str()),
                    is_list,
                    visited,
                ))
            }
        }

        chunks
    }

    fn find_chunks(&self) -> Chunk<Chunk<FieldName<'a>>> {
        match &self.config.schema.query {
            None => Chunk::new(),
            Some(query) => self.iter(
                Chunk::new(),
                TypeName::new(query.as_str()),
                false,
                &mut HashSet::new(),
            ),
        }
    }

    fn find(&self) -> Vec<Vec<&'a str>> {
        self.find_chunks()
            .as_vec()
            .iter()
            .map(|chunk| {
                chunk
                    .as_vec()
                    .iter()
                    .map(|field_name| field_name.as_str())
                    .collect()
            })
            .collect()
    }
}

#[derive(Clone)]
enum Chunk<A> {
    Nil,
    Cons(A, Rc<Chunk<A>>),
    Concat(Rc<Chunk<A>>, Rc<Chunk<A>>),
}

impl<A> Chunk<A> {
    fn new() -> Self {
        Self::Nil
    }

    fn from_vec(vec: Vec<A>) -> Chunk<A> {
        let mut path = Chunk::new();
        for a in vec.into_iter().rev() {
            path = Chunk::Cons(a, Rc::new(path));
        }
        path
    }

    fn append(self, a: A) -> Self {
        Chunk::Cons(a, Rc::new(self))
    }

    fn concat(self, other: Chunk<A>) -> Self {
        Self::Concat(Rc::new(self), Rc::new(other))
    }

    fn as_vec(&self) -> Vec<&A> {
        let mut vec = Vec::new();
        self.as_vec_mut(&mut vec);
        vec
    }

    fn as_vec_mut<'a>(&'a self, buf: &mut Vec<&'a A>) {
        match self {
            Chunk::Nil => {}
            Chunk::Cons(a, rest) => {
                buf.push(a);
                rest.as_vec_mut(buf);
            }
            Chunk::Concat(a, b) => {
                a.as_vec_mut(buf);
                b.as_vec_mut(buf);
            }
        }
    }
}
