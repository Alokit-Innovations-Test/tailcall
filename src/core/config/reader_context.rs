use std::borrow::Cow;
use std::collections::BTreeMap;

use headers::HeaderMap;
use serde::Deserialize;
use crate::core::has_headers::HasHeaders;
use crate::core::json::JsonLike;
use crate::core::path::PathString;
use crate::core::runtime::TargetRuntime;

pub struct ConfigReaderContext<'a, Value> {
    pub runtime: &'a TargetRuntime<Value>,
    pub vars: &'a BTreeMap<String, String>,
    pub headers: HeaderMap,
}

impl<'a, Value: JsonLike<'a> + Deserialize<'a> + Clone> PathString for ConfigReaderContext<'a, Value> {
    fn path_string<T: AsRef<str>>(&self, path: &[T]) -> Option<Cow<'_, str>> {
        if path.is_empty() {
            return None;
        }

        path.split_first()
            .and_then(|(head, tail)| match head.as_ref() {
                "vars" => self.vars.get(tail[0].as_ref()).map(|v| v.into()),
                "env" => self.runtime.env.get(tail[0].as_ref()),
                _ => None,
            })
    }
}

impl<Value> HasHeaders for ConfigReaderContext<'_, Value> {
    fn headers(&self) -> &HeaderMap {
        &self.headers
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use super::*;
    use crate::core::tests::TestEnvIO;

    #[test]
    fn path_string() {
        let mut runtime = crate::core::runtime::test::init(None);
        runtime.env = Arc::new(TestEnvIO::from_iter([(
            "ENV_1".to_owned(),
            "ENV_VAL".to_owned(),
        )]));

        let reader_context = ConfigReaderContext {
            runtime: &runtime,
            vars: &BTreeMap::from_iter([("VAR_1".to_owned(), "VAR_VAL".to_owned())]),
            headers: Default::default(),
        };

        assert_eq!(
            reader_context.path_string(&["env", "ENV_1"]),
            Some("ENV_VAL".into())
        );
        assert_eq!(reader_context.path_string(&["env", "ENV_5"]), None);
        assert_eq!(
            reader_context.path_string(&["vars", "VAR_1"]),
            Some("VAR_VAL".into())
        );
        assert_eq!(reader_context.path_string(&["vars", "VAR_6"]), None);
        assert_eq!(reader_context.path_string(&["unknown", "unknown"]), None);
    }
}
