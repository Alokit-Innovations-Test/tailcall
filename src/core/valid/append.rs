pub trait Append<A> {
    type Out;
    fn append(self, a: A) -> Self::Out;
}

impl<A0, A1> Append<A1> for (A0,) {
    type Out = (A0, A1);
    fn append(self, a1: A1) -> Self::Out {
        let (a0,) = self;
        (a0, a1)
    }
}

impl<A0, A1, A2> Append<A2> for (A0, A1) {
    type Out = (A0, A1, A2);
    fn append(self, a2: A2) -> Self::Out {
        let (a0, a1) = self;
        (a0, a1, a2)
    }
}

impl<A0, A1, A2, A3> Append<A3> for (A0, A1, A2) {
    type Out = (A0, A1, A2, A3);
    fn append(self, a3: A3) -> Self::Out {
        let (a0, a1, a2) = self;
        (a0, a1, a2, a3)
    }
}

impl<A0, A1, A2, A3, A4> Append<A4> for (A0, A1, A2, A3) {
    type Out = (A0, A1, A2, A3, A4);
    fn append(self, a4: A4) -> Self::Out {
        let (a0, a1, a2, a3) = self;
        (a0, a1, a2, a3, a4)
    }
}

impl<A0, A1, A2, A3, A4, A5> Append<A5> for (A0, A1, A2, A3, A4) {
    type Out = (A0, A1, A2, A3, A4, A5);
    fn append(self, a5: A5) -> Self::Out {
        let (a0, a1, a2, a3, a4) = self;
        (a0, a1, a2, a3, a4, a5)
    }
}

impl<A0, A1, A2, A3, A4, A5, A6> Append<A6> for (A0, A1, A2, A3, A4, A5) {
    type Out = (A0, A1, A2, A3, A4, A5, A6);
    fn append(self, a6: A6) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5) = self;
        (a0, a1, a2, a3, a4, a5, a6)
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7> Append<A7> for (A0, A1, A2, A3, A4, A5, A6) {
    type Out = (A0, A1, A2, A3, A4, A5, A6, A7);
    fn append(self, a7: A7) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6) = self;
        (a0, a1, a2, a3, a4, a5, a6, a7)
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8> Append<A8> for (A0, A1, A2, A3, A4, A5, A6, A7) {
    type Out = (A0, A1, A2, A3, A4, A5, A6, A7, A8);
    fn append(self, a8: A8) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7) = self;
        (a0, a1, a2, a3, a4, a5, a6, a7, a8)
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9> Append<A9> for (A0, A1, A2, A3, A4, A5, A6, A7, A8) {
    type Out = (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9);
    fn append(self, a9: A9) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8) = self;
        (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9)
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10> Append<A10>
    for (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9)
{
    type Out = (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10);
    fn append(self, a10: A10) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9) = self;
        (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10)
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11> Append<A11>
    for (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10)
{
    type Out = (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11);
    fn append(self, a11: A11) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10) = self;
        (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11)
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12> Append<A12>
    for (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11)
{
    type Out = (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12);
    fn append(self, a12: A12) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11) = self;
        (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12)
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13> Append<A13>
    for (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12)
{
    type Out = (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13);
    fn append(self, a13: A13) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12) = self;
        (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13)
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14> Append<A14>
    for (A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13)
{
    type Out = (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
    );
    fn append(self, a14: A14) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13) = self;
        (
            a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14,
        )
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15> Append<A15>
    for (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
    )
{
    type Out = (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
    );
    fn append(self, a15: A15) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14) = self;
        (
            a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15,
        )
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15, A16> Append<A16>
    for (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
    )
{
    type Out = (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
    );
    fn append(self, a16: A16) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15) = self;
        (
            a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16,
        )
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15, A16, A17> Append<A17>
    for (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
    )
{
    type Out = (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
        A17,
    );
    fn append(self, a17: A17) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16) = self;
        (
            a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17,
        )
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15, A16, A17, A18>
    Append<A18>
    for (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
        A17,
    )
{
    type Out = (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
        A17,
        A18,
    );
    fn append(self, a18: A18) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17) = self;
        (
            a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17, a18,
        )
    }
}

impl<A0, A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15, A16, A17, A18, A19>
    Append<A19>
    for (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
        A17,
        A18,
    )
{
    type Out = (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
        A17,
        A18,
        A19,
    );
    fn append(self, a19: A19) -> Self::Out {
        let (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17, a18) =
            self;
        (
            a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17, a18,
            a19,
        )
    }
}

impl<
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
        A17,
        A18,
        A19,
        A20,
    > Append<A20>
    for (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
        A17,
        A18,
        A19,
    )
{
    type Out = (
        A0,
        A1,
        A2,
        A3,
        A4,
        A5,
        A6,
        A7,
        A8,
        A9,
        A10,
        A11,
        A12,
        A13,
        A14,
        A15,
        A16,
        A17,
        A18,
        A19,
        A20,
    );
    fn append(self, a20: A20) -> Self::Out {
        let (
            a0,
            a1,
            a2,
            a3,
            a4,
            a5,
            a6,
            a7,
            a8,
            a9,
            a10,
            a11,
            a12,
            a13,
            a14,
            a15,
            a16,
            a17,
            a18,
            a19,
        ) = self;
        (
            a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17, a18,
            a19, a20,
        )
    }
}
