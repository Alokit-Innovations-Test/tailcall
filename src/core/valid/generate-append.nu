#! /usr/bin/env nu

# Generates ./append.rs file for specified N number of possible variants
# If needed change the N and run the script with nushell

const N = 20

mut result = "
pub trait Append<A> {
    type Out;
    fn append(self, a: A) -> Self::Out;
}

impl<A0, A1> Append<A1> for (A0,) {
    type Out = (A0, A1);
    fn append(self, a1: A1) -> Self::Out {
        let (a0,) = self;
        (a0, a1)
    }
}
"

for i in 2..$N {
    let prev = 0..($i - 1) | each {|i| $"A($i)"} | str join ", "
    let current = $"A($i)"
    let next = $prev + ", " + $current

    $result += $"
impl<($next)> Append<($current)> for \(($prev)\) {
    type Out = \(($next)\);
    fn append\(self, ($current | str downcase): ($current)\) -> Self::Out {
        let \(($prev | str downcase)\) = self;
        \(($next | str downcase)\)
    }
}
    "
}

$result | save -f append.rs
