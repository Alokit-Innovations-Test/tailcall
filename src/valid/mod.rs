mod cause;
mod error;
mod valid;
mod zipped;

pub use cause::*;
pub use error::*;
pub use valid::*;
pub use zipped::*;
