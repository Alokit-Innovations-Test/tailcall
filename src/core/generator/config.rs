use std::path::Path;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::config::{self, reader::ConfigReader};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub enum InputSource {
    /// src: maintains the same src written in config.
    /// resolved_src: holds the correctly resolved src with respect to config.
    Config {
        src: String,
        resolved_src: Option<String>,
    },
    Import {
        src: String,
        resolved_src: Option<String>,
    },
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Input {
    #[serde(flatten)]
    pub source: InputSource,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Output {
    /// Controls the output format (graphql, json, yaml)
    pub format: config::Source,
    /// Specifies the output file name
    pub file: String,
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    #[serde(default = "defaults::schema::query")]
    pub query: String,
    #[serde(default = "defaults::schema::mutation")]
    pub mutation: String,
}

impl Default for Schema {
    fn default() -> Self {
        Self {
            query: defaults::schema::query(),
            mutation: defaults::schema::mutation(),
        }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GenerateOptions {
    #[serde(default)]
    pub schema: Schema,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    // TODO: change types
    pub ambiguous_name_resolver: Option<serde_json::Value>,
    pub merge_types: Option<serde_json::Value>,
    pub js: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GeneratorConfig {
    pub input: Vec<Input>,
    #[serde(default)]
    pub output: Output,
    #[serde(default)]
    pub generate: GenerateOptions,
    #[serde(default)]
    pub transform: Transform,
}

impl GeneratorConfig {
    /// Resolves all the paths present inside the GeneratorConfig.
    pub fn resolve_paths(mut self, file_path: &str) -> Self {
        let config_path = Path::new(file_path).parent().unwrap_or(Path::new(""));

        for input in self.input.iter_mut() {
            match &mut input.source {
                InputSource::Config { src, resolved_src }
                | InputSource::Import { src, resolved_src } => {
                    *resolved_src = Some(ConfigReader::resolve_path(src, Some(config_path)));
                }
            }
        }

        self.output.file = ConfigReader::resolve_path(&self.output.file, Some(config_path));

        self
    }
}

mod defaults {
    pub mod schema {
        pub fn query() -> String {
            "Query".to_string()
        }

        pub fn mutation() -> String {
            "Mutation".to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::GeneratorConfig;

    #[test]
    fn test_from_json() {
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();

        assert_debug_snapshot!(&config);
    }

    #[test]
    fn test_resolve_paths() {
        let file_path = "../../../tailcall-fixtures/fixtures/generator/simple-json.json";
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();
        let config = config.resolve_paths(file_path);
        assert_debug_snapshot!(&config);
    }
}
