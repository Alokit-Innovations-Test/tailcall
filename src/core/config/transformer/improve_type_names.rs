use std::collections::HashSet;

use convert_case::{Case, Casing};
use indexmap::IndexMap;

use crate::core::config::Config;
use crate::core::transform::Transform;
use crate::core::valid::Valid;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct CandidateStats {
    frequency: u32,
    priority: Priority,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Priority {
    High,
    Medium,
}

impl Priority {
    fn as_u8(&self) -> u8 {
        match self {
            Priority::High => 3,
            Priority::Medium => 2,
        }
    }
}

impl std::cmp::Ord for Priority {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.as_u8().cmp(&other.as_u8())
    }
}

impl std::cmp::PartialOrd for Priority {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

struct CandidateConvergence<'a> {
    /// maintains the generated candidates in the form of
    /// {TypeName: {{candidate_name: {frequency: 1, priority: 2}}}}
    candidates: IndexMap<String, IndexMap<String, CandidateStats>>,
    config: &'a Config,
}

impl<'a> CandidateConvergence<'a> {
    fn new(candate_gen: CandidateGeneration<'a>) -> Self {
        Self {
            candidates: candate_gen.candidates,
            config: candate_gen.config,
        }
    }

    /// Converges on the most frequent candidate name for each type.
    /// This method selects the most frequent candidate name for each type,
    /// ensuring uniqueness.
    fn converge(self) -> IndexMap<String, String> {
        let mut finalized_candidates = IndexMap::new();
        let mut converged_candidate_set = HashSet::new();

        for (type_name, candidate_list) in self.candidates.iter() {
            // Filter out candidates that have already been converged or are already present
            // in types
            let candidates_to_consider = candidate_list.iter().filter(|(candidate_name, _)| {
                let candidate_type_name = candidate_name.to_case(Case::Pascal);
                !converged_candidate_set.contains(&candidate_type_name)
                    && !self.config.types.contains_key(&candidate_type_name)
            });

            // Find the candidate with the highest frequency and priority
            if let Some((candidate_name, _)) = candidates_to_consider
                .max_by_key(|(key, value)| (value.priority, value.frequency, *key))
            {
                let singularized_candidate_name = candidate_name.to_case(Case::Pascal);
                finalized_candidates
                    .insert(type_name.to_owned(), singularized_candidate_name.clone());
                converged_candidate_set.insert(singularized_candidate_name);
            }
        }

        finalized_candidates
    }
}

struct CandidateGeneration<'a> {
    /// maintains the generated candidates in the form of
    /// {TypeName: {{candidate_name: {frequency: 1, priority: 2}}}}
    candidates: IndexMap<String, IndexMap<String, CandidateStats>>,
    config: &'a Config,
}

impl<'a> CandidateGeneration<'a> {
    fn new(config: &'a Config) -> Self {
        Self { candidates: Default::default(), config }
    }

    /// Generates candidate type names based on the provided configuration.
    /// This method iterates over the configuration and collects candidate type
    /// names for each type.
    fn generate(mut self) -> CandidateConvergence<'a> {
        let mut types_processing_order = [
            &self.config.schema.query,
            &self.config.schema.mutation,
            &self.config.schema.subscription,
        ]
        .iter()
        .flat_map(|t| t.as_ref())
        .chain(self.config.types.keys())
        .collect::<Vec<_>>();

        // we want to process the operation types first so we add it manually and then
        // use the all types in config, so inorder to process each type only once we
        // remove the duplicate operation types which got added via
        // `self.config.types.keys()`.
        types_processing_order.dedup();

        for type_name in self.config.types.keys() {
            if let Some(type_info) = self.config.types.get(type_name) {
                for (field_name, field_info) in type_info.fields.iter() {
                    if self.config.is_scalar(&field_info.type_of) {
                        // if output type is scalar then skip it.
                        continue;
                    }

                    let inner_map = self
                        .candidates
                        .entry(field_info.type_of.to_owned())
                        .or_default();

                    let singularized_candidate = pluralizer::pluralize(field_name, 1, false);
                    let priority = if self.config.is_root_operation_type(type_name) {
                        Priority::High // user suggested name has the highest
                                       // priority over
                                       // auto inferred names.
                    } else {
                        Priority::Medium
                    };

                    if let Some(val) = inner_map.get_mut(&singularized_candidate) {
                        val.frequency += 1;
                        val.priority = std::cmp::max(val.priority, priority);
                    } else {
                        // generate the backup name:
                        inner_map.insert(
                            singularized_candidate,
                            CandidateStats { frequency: 1, priority },
                        );
                    }
                }
            }
        }
        CandidateConvergence::new(self)
    }
}

#[derive(Default)]
pub struct ImproveTypeNames;

impl ImproveTypeNames {
    /// Generates type names based on inferred candidates from the provided
    /// configuration.
    fn generate_type_names(&self, mut config: Config) -> Config {
        let finalized_candidates = CandidateGeneration::new(&config).generate().converge();

        for (old_type_name, new_type_name) in finalized_candidates {
            if let Some(type_) = config.types.remove(old_type_name.as_str()) {
                // Add newly generated type.
                config.types.insert(new_type_name.to_owned(), type_);

                // Replace all the instances of old name in config.
                for actual_type in config.types.values_mut() {
                    for actual_field in actual_type.fields.values_mut() {
                        if actual_field.type_of == old_type_name {
                            // Update the field's type with the new name
                            actual_field.type_of.clone_from(&new_type_name);
                        }
                    }
                }
            }
        }
        config
    }
}

impl Transform for ImproveTypeNames {
    type Value = Config;
    type Error = String;
    fn transform(&self, config: Config) -> Valid<Self::Value, Self::Error> {
        let config = self.generate_type_names(config);

        Valid::succeed(config)
    }
}

#[cfg(test)]
mod test {
    use std::fs;

    use anyhow::Ok;
    use tailcall_fixtures::configs;

    use super::ImproveTypeNames;
    use crate::core::config::Config;
    use crate::core::transform::Transform;
    use crate::core::valid::Validator;

    fn read_fixture(path: &str) -> String {
        fs::read_to_string(path).unwrap()
    }

    #[test]
    fn test_type_name_generator_transform() {
        let config = Config::from_sdl(read_fixture(configs::AUTO_GENERATE_CONFIG).as_str())
            .to_result()
            .unwrap();

        let transformed_config = ImproveTypeNames.transform(config).to_result().unwrap();
        insta::assert_snapshot!(transformed_config.to_sdl());
    }

    #[test]
    fn test_type_name_generator_with_cyclic_types() -> anyhow::Result<()> {
        let config = Config::from_sdl(read_fixture(configs::CYCLIC_CONFIG).as_str())
            .to_result()
            .unwrap();

        let transformed_config = ImproveTypeNames.transform(config).to_result().unwrap();
        insta::assert_snapshot!(transformed_config.to_sdl());

        Ok(())
    }

    #[test]
    fn test_type_name_generator() -> anyhow::Result<()> {
        let config = Config::from_sdl(read_fixture(configs::NAME_GENERATION).as_str())
            .to_result()
            .unwrap();

        let transformed_config = ImproveTypeNames.transform(config).to_result().unwrap();
        insta::assert_snapshot!(transformed_config.to_sdl());

        Ok(())
    }

    #[test]
    fn test_prioritize_user_given_names() {
        let sdl = r#"
            schema { query: Query }
            type Post {
                id: Int
            }
            type T2 {
                userPosts: [Posts]
            }
            type Query {
                userInfo: T2
            }
        "#;

        let config = Config::from_sdl(sdl).to_result().unwrap();

        let config = ImproveTypeNames.transform(config).to_result().unwrap();

        insta::assert_snapshot!(config.to_sdl());
    }
}
