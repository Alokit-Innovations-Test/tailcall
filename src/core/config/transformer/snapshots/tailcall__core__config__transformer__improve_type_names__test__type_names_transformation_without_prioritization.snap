---
source: src/core/config/transformer/improve_type_names.rs
expression: transformed_config.to_sdl()
---
schema @server(port: 8000) @upstream(baseURL: "http://jsonplaceholder.typicode.com", batch: {delay: 100, headers: []}, httpCache: 42) {
  query: Query
}

type Query {
  post(id: ID!): TodoPost @http(path: "/posts")
  todos: [Todo] @http(path: "/todos")
  users: [User] @http(path: "/users")
}

type Todo {
  id: ID
  todoPost: TodoPost
}

type TodoPost {
  body: String
  id: ID
  title: String
}

type User {
  id: ID
  name: String
  relatedPosts: [TodoPost] @http(path: "/posts", query: [{key: "id", value: "{{.value.id}}"}])
}
